admin-tools/data_mover.sh

```diff
@@ -0,0 +1,116 @@
+#! /bin/bash
+# Copyright 2014 Telefonica Investigacion y Desarrollo, S.A.U
+#
+# This file is part of fiware-cosmos.
+#
+# fiware-cosmos is free software: you can redistribute it and/or
+# modify it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# fiware-cosmos is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero
+# General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with fiware-cosmos. If not, see http://www.gnu.org/licenses/.
+#
+# For those usages not covered by this license please contact with
+# francisco dot romerobueno at telefonica dot com
+
+# Show the usage
+if [ $# -ne 4 ]; then
+	echo "Usage: ./data-mover.sh <usernames_file> <src_cluster_URL> <dst_cluster_URL> <max_bytes_transfer>"
+	exit 1;
+fi
+
+# Arguments
+usernamesFile=$1
+srcClusterURL=$2
+dstClusterURL=$3
+maxBytesTransfer=$4
+
+# Create a temporal working directory
+tmpDir=$(mktemp -d /tmp/data_mover.XXXX)
+
+# Migrates a given path
+# $1 : path to be migrated
+# $2 : username owning of the path
+migrate() {
+	local path=$1
+	local username=$2
+	local index=1
+
+	# Iterate on the entries of the path
+	while read -r entry; do
+		# Check the type of the entry by getting its size (0 bytes for directories)
+		length=$(curl -X GET "$srcClusterURL/webhdfs/v1$path?op=liststatus&user.name=hdfs" 2> /dev/null | \
+			python -m json.tool | grep length | awk '{ print $2 }' | sed 's/"//g' | sed 's/,//g' | sed $index'q;d')
+
+		if test $length -eq 0; then
+			echo "$srcClusterURL/webhdfs/v1$path/$entry is a directory"
+			# Create the directory in the destination cluster with appropriate ownership
+			curl -X PUT "$dstClusterURL/webhdfs/v1$path/$entry?op=mkdirs&user.name=hdfs" 2> /dev/null > /dev/null
+			curl -X PUT "$dstClusterURL/webhdfs/v1$path/$entry?op=setowner&owner=$username&group=$username&user.name=hdfs" \
+				2> /dev/null > /dev/null
+			# This directory must be migrated as well in a recursive fashion
+			migrate $path/$entry $username
+		else
+			echo "$srcClusterURL/webhdfs/v1$path/$entry is a file ($length bytes)"
+			# Create a temporal file for downloading the entire HDFS file
+			tmpFile=$(mktemp $tmpDir/file.XXXX)
+			# Download the HDFS file
+			curl -X GET "$srcClusterURL/webhdfs/v1$path/$entry?op=open&user.name=hdfs" > $tmpFile 2> /dev/null
+			# Get the number of chunks according to the maximum number of bytes allowed for transferring
+			numChunks=$(($length/$maxBytesTransfer))
+
+			# Copy each chunk into a HDFS file in the destination cluster
+			for i in `seq 0 $numChunks`; do
```

Comment:
Yes. That is because numChunks is about chunks of maxBytesTransfer size. Thus, it is necessary an addition loop for the remaining bytes. E.g. length=12340, maxBytesTransfer=2000 --> numChunks=6, thus the number of loops is 7, 6 for 12000 bytes, and the 7th for 340 bytes.